core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooCache.java

```diff
@@ -152,53 +184,78 @@ public void process(WatchedEvent event) {
   }
 
-  /**
-   * Creates a new cache. The given watcher is called whenever a watched node changes.
-   *
-   * @param zk the ZooKeeper instance
-   * @param watcher watcher object
-   * @throws NullPointerException if zk or watcher is {@code null}
-   */
-  public ZooCache(ZooSession zk, ZooCacheWatcher watcher) {
-    this(zk, Optional.of(watcher), Duration.ofMinutes(3));
+  // for tests that use a Ticker
+  public ZooCache(ZooSession zk, Set<String> pathsToWatch, Ticker ticker) {
+    this.zk = requireNonNull(zk);
+    this.cache = Caches.getInstance().createNewBuilder(Caches.CacheName.ZOO_CACHE, false)
+        .ticker(requireNonNull(ticker)).expireAfterAccess(CACHE_DURATION).build();
+    this.nodeCache = cache.asMap();
+    this.pathsToWatch = requireNonNull(pathsToWatch);
+    setupWatchers(pathsToWatch);
+    log.trace("{} created new cache", cacheId, new Exception());
   }
 
-  public ZooCache(ZooSession zk, Optional<ZooCacheWatcher> watcher, Duration timeout) {
-    this.zk = requireNonNull(zk);
-    this.externalWatcher = watcher;
-    RemovalListener<String,ZcNode> removalListerner = (path, zcNode, reason) -> {
-      try {
-        log.trace("{} removing watches for {} because {} accesses {}", cacheId, path, reason,
-            zcNode == null ? -1 : zcNode.getAccessCount());
-        zk.removeWatches(path, ZooCache.this.watcher, Watcher.WatcherType.Any, false);
-      } catch (InterruptedException | KeeperException | RuntimeException e) {
-        log.warn("{} failed to remove watches on path {} in zookeeper", cacheId, path, e);
+  public void addZooCacheWatcher(ZooCacheWatcher watcher) {
+    externalWatchers.add(requireNonNull(watcher));
+  }
+
+  // Visible for testing
+  protected void setupWatchers(Set<String> pathsToWatch) {
+
+    for (String left : pathsToWatch) {
+      for (String right : pathsToWatch) {
+        if (!left.equals(right) && left.contains(right)) {
+          throw new IllegalArgumentException(
+              "Overlapping paths found in paths to watch. left: " + left + ", right: " + right);
+        }
       }
-    };
-    // Must register the removal listener using evictionListener inorder for removal to be mutually
-    // exclusive with any other operations on the same path. This is important for watcher
-    // consistency, concurrently adding and removing watches for the same path would leave zoocache
-    // in a really bad state. The cache builder has another way to register a removal listener that
-    // is not mutually exclusive.
-    Cache<String,ZcNode> cache =
-        Caches.getInstance().createNewBuilder(Caches.CacheName.ZOO_CACHE, false)
-            .expireAfterAccess(timeout).evictionListener(removalListerner).build();
-    nodeCache = cache.asMap();
-    log.trace("{} created new cache", cacheId, new Exception());
+    }
+
+    try {
+      for (String path : pathsToWatch) {
+        watchedPaths.add(path);
+        zk.addPersistentRecursiveWatcher(path, this.watcher);
```

Comment:
If the path is added to watchedPaths after then in the case of an excpetion it will never be added, that would cause `ensureWatched` calls to fail for this substree.

```suggestion
        zk.addPersistentRecursiveWatcher(path, this.watcher);
        watchedPaths.add(path);
```